### PR TITLE
feat: partial message created as named definitions

### LINKED ISSUE #3706: Generated swagger creates anonymous structure for post methods with path params
## 🐛 Bug Report

For RPC that have a body (typically POST) and one or several path params, the swagger generated uses an anonymous inline schema.
This creates issues with some code generator later on, when generating clients from the swagger definitions

(Arguably the output of protoc-gen-openapiv2 is correct per the spec. What's described here is just very impractical downstream and I think it would be improved by the suggested change)

This is actually visible in this other issue I found https://github.com/grpc-ecosystem/grpc-gateway/issues/2681

## To Reproduce

Take 2 RPCs that use the same input params. But one uses all of them in the body, while the other promotes one to the path

```proto
syntax = "proto3";

import "google/api/annotations.proto";

option go_package = "./test";

service Test{
    rpc AMethod (AMessage) returns (SomeResponse) {
        option (google.api.http) = {
            post: "/v1/projects/someotherpath"
            body: "*"
        };
    }
    rpc BMethod (AMessage) returns (SomeResponse) {
        option (google.api.http) = {
            post: "/v1/projects/someotherpath/{project_id}"
            body: "*"
        };
    }
}
message AMessage {
    string project_id = 1;
    string other_field = 2;
}
message SomeResponse {
    string msg = 1;
}
```

When generating the corresponding swagger
```bash
protoc --plugin=$GOPATH/bin/protoc-gen-openapi --openapiv2_out . --proto_path=$PWD test.proto
```

you get

```json
{
  "swagger": "2.0",
  "info": {
    "title": "test.proto",
    "version": "version not set"
  },
  "tags": [
    {
      "name": "Test"
    }
  ],
  "consumes": [
    "application/json"
  ],
  "produces": [
    "application/json"
  ],
  "paths": {
    "/v1/projects/someotherpath": {
      "post": {
        "operationId": "Test_AMethod",
        "responses": {
          "200": {
            "description": "A successful response.",
            "schema": {
              "$ref": "#/definitions/SomeResponse"
            }
          },
          "default": {
            "description": "An unexpected error response.",
            "schema": {
              "$ref": "#/definitions/rpcStatus"
            }
          }
        },
        "parameters": [
          {
            "name": "body",
            "in": "body",
            "required": true,
            "schema": {
              "$ref": "#/definitions/AMessage"
            }
          }
        ],
        "tags": [
          "Test"
        ]
      }
    },
    "/v1/projects/someotherpath/{projectId}": {
      "post": {
        "operationId": "Test_BMethod",
        "responses": {
          "200": {
            "description": "A successful response.",
            "schema": {
              "$ref": "#/definitions/SomeResponse"
            }
          },
          "default": {
            "description": "An unexpected error response.",
            "schema": {
              "$ref": "#/definitions/rpcStatus"
            }
          }
        },
        "parameters": [
          {
            "name": "projectId",
            "in": "path",
            "required": true,
            "type": "string"
          },
          {
            "name": "body",
            "in": "body",
            "required": true,
            "schema": {
              "type": "object",
              "properties": {
                "otherField": {
                  "type": "string"
                }
              }
            }
          }
        ],
        "tags": [
          "Test"
        ]
      }
    }
  },
  "definitions": {
    "AMessage": {
      "type": "object",
      "properties": {
        "projectId": {
          "type": "string"
        },
        "otherField": {
          "type": "string"
        }
      }
    },
    "SomeResponse": {
      "type": "object",
      "properties": {
        "msg": {
          "type": "string"
        }
      }
    },
    "protobufAny": {
      "type": "object",
      "properties": {
        "@type": {
          "type": "string"
        }
      },
      "additionalProperties": {}
    },
    "rpcStatus": {
      "type": "object",
      "properties": {
        "code": {
          "type": "integer",
          "format": "int32"
        },
        "message": {
          "type": "string"
        },
        "details": {
          "type": "array",
          "items": {
            "type": "object",
            "$ref": "#/definitions/protobufAny"
          }
        }
      }
    }
  }
}

```

What you can see, is that because `AMethod` uses the entire `AMessage` in the body, then the reported schema is
```json
"parameters": [
          {
            "name": "body",
            "in": "body",
            "required": true,
            "schema": {
              "$ref": "#/definitions/AMessage"
            }
          }
        ],
```
which references the proto defined `AMessage` object

However, for `BMethod`, the input proto message is split between the path and the body. This ends up with the swagger using only inline definitions of params (I guess because there is no matching proto-defined object in the definitions)

```json
"parameters": [
          {
            "name": "projectId",
            "in": "path",
            "required": true,
            "type": "string"
          },
          {
            "name": "body",
            "in": "body",
            "required": true,
            "schema": {
              "type": "object",
              "properties": {
                "otherField": {
                  "type": "string"
                }
              }
            }
          }
        ],
```

This, in turn, with some code generator like `./node_modules/.bin/openapi-generator-cli` generates code that looks like this
```ts
public bMethod(projectId: string, body: InlineObject3, observe?: 'body', reportProgress?: boolean, options?: {httpHeaderAccept?: 'application/json',}): Observable<object>;
```

you'll note the use of `InlineObject3` which is not really something we can build upon as this name is based on the order of generation, which might change as the proto evolve

## Expected behavior

The generated swagger should not define body schema inline, but should always used reference to named objects in the `definitions` section

We could use some naming convention like `BMethodBody`. This is what the `go-swagger` code generator actually does it seems.

## Actual Behavior

(Write what happened. Add screenshots, if applicable.)

## Your Environment

```
libprotoc 24.3
PROTOC_GEN_GRPC_GATEWAY_VERSION=v2.18.0
```

### PR DESCRIPTION
**Pull request by gbataille**
cc @johanbrandhorst
cc @jarnault

#### References to other Issues or PRs

Fixes #3706 

#### Brief description of what is fixed or changed

Useful for proto Message that are used in a POST method for example where the properties of the proto message are split between path params and body params:
For downstream code generator, you want the body params to be wrapped inside a **named** swagger definitions object

#### Other comments

DRAFT:
There are still plenty of traces in there but I'm not done. I have tested it locally and it seems to produce the right swagger, but I'd like my front end developer colleague to try to generate his TS code with his toolset to see if it behaves as expected.

Once I have this testing done I'll clean that PR

TEST:
I did not find any unit tests that includes the `renderServices` part or the `applyTemplate` and I guess I would have a hard time to create some from scratch

QUESTIONS:
- I worked with the assumption that a method name is unique, which I think is true. Adding a `Body` suffix to make a definition object on the fly seems sufficient, do you agree?
- I think a better way would be to have been able to add the custom object to the registry to ask for it to be added later (through `customRefs`) but I did not find a way to add things to the registry, therefore the way the definition is added is manual and the way the schema `ref` is named is hardcoded. I did not find anything better. Any idea?

RESULT:

Before, with a proto like 
```proto
service Test{
    rpc AMethod (AMessage) returns (SomeResponse) {
        option (google.api.http) = {
            post: "/v1/projects/someotherpath"
            body: "*"
        };
    }
    rpc BMethod (AMessage) returns (SomeResponse) {
        option (google.api.http) = {
            post: "/v1/projects/someotherpath/{project_id}"
            body: "*"
        };
    }
}
message AMessage {
    string project_id = 1;
    string other_field = 2;
}
message SomeResponse {
    string msg = 1;
}
```

you would get a swagger like that
```json
    "/v1/projects/someotherpath/{projectId}": {
      "post": {
        "operationId": "Test_BMethod",
        "responses": {
          "200": {
            "description": "A successful response.",
            "schema": {
              "$ref": "#/definitions/SomeResponse"
            }
          },
          "default": {
            "description": "An unexpected error response.",
            "schema": {
              "$ref": "#/definitions/rpcStatus"
            }
          }
        },
        "parameters": [
          {
            "name": "projectId",
            "in": "path",
            "required": true,
            "type": "string"
          },
          {
            "name": "body",
            "in": "body",
            "required": true,
            "schema": {
              "type": "object",
              "properties": {
                "otherField": {
                  "type": "string"
                }
              }
            }
          }
        ],
        "tags": [
          "Test"
        ]
      }
    }
  },

```

Now you get
```json
    "/v1/projects/someotherpath/{projectId}": {
      "post": {
        [...]
        "parameters": [
          {
            "name": "projectId",
            "in": "path",
            "required": true,
            "type": "string"
          },
          {
            "name": "body",
            "in": "body",
            "required": true,
            "schema": {
              "$ref": "#/definitions/BMethodBody"
            }
          }
        ],
        "tags": [
          "Test"
        ]
      }
    }
  },
  "definitions": {
[...]
    "BMethodBody": {
      "type": "object",
      "properties": {
        "otherField": {
          "type": "string"
        }
      }
    },
[...]
```

## Summary by CodeRabbit

- **Refactor**
	- Enhanced the service rendering function to improve path parameters, body, and path construction handling.
- **New Features**
	- Added additional logging for better visibility and debugging.

### CODE REVIEW

**Comment by gbataille** [comment]
Hey @johanbrandhorst 

so the good news is that on our real application (as opposed to the toy example I was playing with), it works.
with one caveat - there is some kind of namespacing that is missing. I need to have a look into that.
I'll try to include it and remove all the log traces.

Any comment on the code / strategy? As mentioned, I'm not super happy but I did not find any better solution

File: protoc-gen-openapiv2/internal/genopenapi/template.go [hunk 1]
```diff
@@ -1129,11 +1129,13 @@ func renderServices(services []*descriptor.Service, paths *openapiPathsObject, r
 		}
 
 		for methIdx, meth := range svc.Methods {
+			grpclog.Errorf("METHOD:\n%+v\n\n", meth)
 			if !isVisible(getMethodVisibilityOption(meth), reg) {
 				continue
 			}
 
 			for bIdx, b := range meth.Bindings {
+				grpclog.Errorf("BINDINGS:\n%v\n\n", b)
```
[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply protoc-gen-openapiv2/internal/genopenapi/template.go:1138]: I assume these are debugging logs, do you mean to remove them?

File: protoc-gen-openapiv2/internal/genopenapi/template.go [hunk 2]
```diff
@@ -1276,15 +1280,26 @@ func renderServices(services []*descriptor.Service, paths *openapiPathsObject, r
...
 									return err
 								}
 								desc = messageSchema.Description
 							} else {
-								schema = messageSchema
-								if schema.Properties == nil || len(*schema.Properties) == 0 {
-									grpclog.Warningf("created a body with 0 properties in the message, this might be unintended: %s", *meth.RequestType)
+								if meth.Name != nil {
+									grpclog.Errorf("method name:\n%v\n\n", *meth.Name)
+									defName := fmt.Sprintf("%sBody", *meth.Name)
```
Comment (johanbrandhorst) [inline protoc-gen-openapiv2/internal/genopenapi/template.go:1295]: We'll need to use more than just this, probably including the protobuf file path (with slashes replaced by `_` perhaps?).

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]
Thanks for this PR, looks like a great start. I would add that we'll need to be more verbose with the namespacing, as you suggest. Remember that the generator supports merging multiple protobuf files into a single swagger definition, so I think we'll want to use the file name (which, relative to the protobuf root, must be unique), service name and method name, all appended. I don't know that the name matters all that much, as long as it's guaranteed to be unique, right?

As for tests, I think you should be able to add a new test to the template_test.go file. All the tests in there test the generator logic. You're right that there isn't one for the `renderServices` function specifically, so you'll either have to create that or just use one of the existing tests as a template for a new test. It shouldn't be too bad to throw together a test with path parameters that affect the final rendered contents?

I think you know this already, but you'll need to run the generation script too, I expect there will be changes to existing generated files.

**Comment by gbataille** [comment]
Hey @johanbrandhorst 

yes as I mentioned and you point out, this was not over and I wanted "real-life" tests. that's where I realized the namespacing issue.

Now due to completely unexpected life-change for me, I won't be able to finish this work. I'm very sorry
@jarnault I guess that's your chance to do some go if you want!

**Comment by johanbrandhorst** [comment]
Closing in favor of #3743

**Pull request closed without merging**